flink-runtime/src/test/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPoolTest.java

```diff
@@ -0,0 +1,286 @@
+            }
+
+            // recycle one by one
+            for (MemorySegment segment : buffers.remove(owners[0])) {
+                bufferPool.recycle(segment, owners[0]);
+            }
+
+            // bulk recycle
+            bufferPool.recycle(buffers.remove(owners[1]), owners[1]);
+
+            for (Thread requestThread : requestThreads) {
+                requestThread.join();
+            }
+
+            assertNull(exception.get());
+            assertEquals(0, bufferPool.getAvailableBuffers());
+            assertEquals(4, buffers.size());
+        } finally {
+            for (Object owner : buffers.keySet()) {
+                bufferPool.recycle(buffers.remove(owner), owner);
+            }
+            assertEquals(bufferPool.getNumTotalBuffers(), bufferPool.getAvailableBuffers());
+            bufferPool.destroy();
+        }
+    }
+
+    @Test
+    public void testMultipleThreadRequestAndRecycle() throws Exception {
+        int numRequestThreads = 10;
+        AtomicReference<Throwable> exception = new AtomicReference<>();
+        BatchShuffleReadBufferPool bufferPool = createBufferPool();
+
+        try {
+            Thread[] requestThreads = new Thread[numRequestThreads];
+            for (int i = 0; i < numRequestThreads; ++i) {
+                requestThreads[i] =
+                        new Thread(
+                                () -> {
+                                    try {
+                                        Object owner = new Object();
+                                        for (int j = 0; j < 100; ++j) {
+                                            List<MemorySegment> buffers =
+                                                    bufferPool.requestBuffers(owner);
+                                            Thread.sleep(10);
+                                            if (j % 2 == 0) {
+                                                bufferPool.recycle(buffers, owner);
+                                            } else {
+                                                for (MemorySegment segment : buffers) {
+                                                    bufferPool.recycle(segment, owner);
+                                                }
+                                            }
+                                        }
+                                    } catch (Throwable throwable) {
+                                        exception.set(throwable);
+                                    }
+                                });
+                requestThreads[i].start();
+            }
+
+            for (Thread requestThread : requestThreads) {
+                requestThread.join();
+            }
+
+            assertNull(exception.get());
+            assertEquals(bufferPool.getNumTotalBuffers(), bufferPool.getAvailableBuffers());
+        } finally {
+            bufferPool.destroy();
+        }
+    }
+
+    @Test(expected = IllegalStateException.class)
```

Comment:
could we have this test case to test normal case only and have a separate test case `testRequestBuffersAfterPoolDestroyed` to test `requestBuffers` after destroy?